include/sys/async_notify.h

```diff
@@ -0,0 +1,345 @@
+ * service-specific, but the value shall be non-negative if the
+ * operation succeeded, and negative if the operation failed.
+ *
+ * @return If the notification is to be done by callback this returns
+ * the generic version of the function to be invoked.  The caller must
+ * immediately invoke that function with whatever arguments are
+ * expected by the callback.  If notification is by spin-wait or
+ * signal, the notification has been completed by the point this
+ * function returns.
+ */
+async_notify_generic_callback async_notify_finalize(struct async_notify *notify,
+						    int res);
+
+/**
+ * @brief Check for and read the result of an asynchronous operation.
+ *
+ * @param notify pointer to the object used to specify asynchronous
+ * function behavior and store completion information.
+ *
+ * @param result pointer to storage for the result of the operation.
+ * The result is stored only if the operation has completed.
+ *
+ * @retval 0 if the operation has completed.
+ * @retval -EAGAIN if the operation has not completed.
+ */
+static inline int async_notify_fetch_result(const struct async_notify *notify,
+					    int *result)
+{
+	__ASSERT_NO_MSG(notify != NULL);
+	__ASSERT_NO_MSG(result != NULL);
+	int rv = -EAGAIN;
+
+	if (async_notify_get_method(notify) == ASYNC_NOTIFY_METHOD_COMPLETED) {
+		rv = 0;
+		*result = notify->result;
+	}
+	return rv;
+}
+
+/**
+ * @brief Initialize a notify object for spin-wait notification.
+ *
+ * Clients that use this initialization receive no asynchronous
+ * notification, and instead must periodically check for completion
+ * using async_notify_fetch_result().
+ *
+ * On completion of the operation the client object must be
+ * reinitialized before it can be re-used.
+ *
+ * @param notify pointer to the notification configuration object.
+ */
+static inline void async_notify_init_spinwait(struct async_notify *notify)
+{
+	__ASSERT_NO_MSG(notify != NULL);
+
+	*notify = (struct async_notify){
+		.flags = ASYNC_NOTIFY_METHOD_SPINWAIT,
+	};
+}
+
+/**
+ * @brief Initialize a notify object for (k_poll) signal notification.
+ *
+ * Clients that use this initialization will be notified of the
+ * completion of operations through the provided signal.
+ *
+ * On completion of the operation the client object must be
+ * reinitialized before it can be re-used.
+ *
+ * @note
+ *   @rst
+ *   This capability is available only when :option:`CONFIG_POLL` is
+ *   selected.
+ *   @endrst
+ *
+ * @param notify pointer to the notification configuration object.
+ *
+ * @param sigp pointer to the signal to use for notification.  The
+ * value must not be null.  The signal must be reset before the client
+ * object is passed to the on-off service API.
+ */
+static inline void async_notify_init_signal(struct async_notify *notify,
+					    struct k_poll_signal *sigp)
+{
+	__ASSERT_NO_MSG(notify != NULL);
+	__ASSERT_NO_MSG(sigp != NULL);
+
+	*notify = (struct async_notify){
+#ifdef CONFIG_POLL
+		.method = {
```

Comment:
Without the `#ifdef` the header won't compile when `CONFIG_POLL=n` because it references a field that isn't defined in that case.  It should be possible to use async notify without enabling poll, so in this case the problem would be diagnosed at runtime by refusing to accept the request.